servicetalk-http-api/src/main/java/io/servicetalk/http/api/ContentCodec.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.api;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+import io.servicetalk.concurrent.api.Publisher;
+
+/**
+ * Codec used to encode and decode {@link Publisher} of {@link Buffer} parts.
+ * This instance is shared therefore it must provide thread safety semantics.
+ */
+public interface ContentCodec {
```

Comment:
A few times yes, I played with having the two together. However, at the end I chose to keep them separate. There is one fundamental difference between the Message & Content codecs, and that's their lifecycle. The former is a lifecycle per message, but the latter maintains context of the encoders for the whole payload. There are some code similarities on the aggregated signatures, but its not that much to worth keeping them in one place for now.

Reply:
I think we can merge streaming and non-streaming variants in a single interface and make the streaming variant in the way it applies the non-streaming method as a mapping function for each chunk. Some compression algorithms work in this way, like LZ4. It will reduce the need for implementing a streaming variant when it's not necessary, like for grpc.

Another way is to provide `ContentCodec` and `StreamingContentCodec` interfaces. The main idea is to make the compression/decompression logic protocol agnostic, without embedding the logic that manages headers, for cases if we have more protocols in the future. That way users will be able to reuse the codec algorithms without implementing a custom codec for each protocol.

Reply:
This discussion moves to https://github.com/apple/servicetalk/pull/1198#discussion_r518407988